agent/utils/utils_test.go

```diff
@@ -219,6 +220,196 @@ func TestBufferedReadSeeker(t *testing.T) {
+				w.Header().Add("Connection", "FooBar")
+				w.Header().Add("Trailer", "Foo")
+				w.Write([]byte("OK"))
+				w.Header().Add("Foo", "Bar")
+			}),
+			WantResponse: &http.Response{
+				Status:     http.StatusText(http.StatusOK),
+				StatusCode: http.StatusOK,
+				Header: http.Header{
+					"A": []string{"B"},
+					"C": []string{"D"},
+				},
+				Trailer: http.Header{
+					"Foo": []string{"Bar"},
+				},
+			},
+			WantResponseBody: "OK",
+		},
+		{
+			Description: "Filter hop trailers",
+			Request:     &http.Request{},
+			Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Add("Trailer", "Foo")
+				w.Header().Add("Trailer", "Connection")
+				w.Write([]byte("OK"))
+				w.Header().Add("Foo", "Bar")
+				w.Header().Add("Connection", "FooBar")
+				w.Header().Add("Trailer:Baz", "Bat")
+				w.Header().Add("Trailer:Upgrade", "BarFoo")
+			}),
+			WantResponse: &http.Response{
+				Status:     http.StatusText(http.StatusOK),
+				StatusCode: http.StatusOK,
+				Header:     http.Header{},
+				Trailer: http.Header{
+					"Foo": []string{"Bar"},
+					"Baz": []string{"Bat"},
+				},
+			},
+			WantResponseBody: "OK",
+		},
+		{
+			Description: "Error response",
+			Request:     &http.Request{},
+			Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Add("Trailer", "Foo")
+				w.Header().Add("Trailer", "Connection")
+				http.Error(w, "test error", http.StatusInternalServerError)
+				w.Header().Add("Foo", "Bar")
+				w.Header().Add("Connection", "FooBar")
+				w.Header().Add("Trailer:Baz", "Bat")
+				w.Header().Add("Trailer:Upgrade", "BarFoo")
+			}),
+			WantResponse: &http.Response{
+				Status:     http.StatusText(http.StatusInternalServerError),
+				StatusCode: http.StatusInternalServerError,
+				Header: http.Header{
+					"Content-Type":           {"text/plain; charset=utf-8"},
+					"X-Content-Type-Options": {"nosniff"},
+				},
+				Trailer: http.Header{
+					"Foo": []string{"Bar"},
+					"Baz": []string{"Bat"},
+				},
+			},
+			WantResponseBody: "test error\n",
+		},
+	}
+	for _, testCase := range testCases {
+		respChan := make(chan *http.Response, 1)
+		rw := NewStreamingResponseWriter(respChan, testCase.Request)
+		go func() {
+			testCase.Handler.ServeHTTP(rw, testCase.Request)
+			rw.Close()
+		}()
+		resp := <-respChan
+		defer resp.Body.Close()
+		if diff := cmp.Diff(resp, testCase.WantResponse, cmpopts.IgnoreFields(http.Response{}, "Body", "Trailer")); len(diff) > 0 {
+			t.Errorf("Unexpected diff in response for %q: %s", testCase.Description, diff)
+		}
+		gotBody, err := io.ReadAll(resp.Body)
+		if err != nil {
+			t.Errorf("Failure reading the response body for %q: %+v", testCase.Description, err)
+		} else if got, want := string(gotBody), testCase.WantResponseBody; got != want {
+			t.Errorf("Unexpected response body for %q: got %q, want %q", testCase.Description, got, want)
+		}
+		if len(testCase.WantResponse.Trailer) > 0 {
```

Comment:
I can see that the trailer verification is not very strict - it doesn't check for unexpected trailers, or values beyond the first one. But I guess checking those would be excessive?